packages/jupyter-ai/jupyter_ai/actors.py

```diff
@@ -0,0 +1,185 @@
+        
+        actor.process_message.remote(message)
+
+
+@ray.remote
+class DocumentIndexActor():
+    def __init__(self, reply_queue: Queue, root_dir: str, log):
+        self.reply_queue = reply_queue
+        self.root_dir = root_dir
+        self.log = log
+        self.index_save_dir = os.path.join(jupyter_data_dir(), '.jupyter_ai_indexes')
+
+        if ChatOpenAINewProvider.auth_strategy.name not in os.environ:
+            return
+        
+        if not os.path.exists(self.index_save_dir):
+            os.makedirs(self.index_save_dir)
+
+        embeddings = OpenAIEmbeddings()
+        try:
+            self.index = FAISS.load_local(self.index_save_dir, embeddings)
+        except Exception as e:
+            self.index = FAISS.from_texts(["This is just starter text for the index"], embeddings)
+
+    def get_index(self):
+        return self.index
+
+    def process_message(self, message: HumanChatMessage):
+        dir_path = message.body.split(' ', 1)[-1]
+        load_path = os.path.join(self.root_dir, dir_path)
+        loader = DirectoryLoader(
+            load_path, 
+            glob="**/*.txt",
+            loader_cls=TextLoader
+        )
+        documents = loader.load_and_split(
+            text_splitter=RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=20)
+        )
+        self.index.add_documents(documents)
+        self.index.save_local(self.index_save_dir)
+
+        response = f"""🎉 I have read documents at **{dir_path}** and ready to answer questions. 
+        You can ask questions from these docs by prefixing your message with **/fs**."""
+        agent_message = AgentChatMessage(
+                id=uuid4().hex,
+                time=time.time(),
+                body=response,
+                reply_to=message.id
+            )
+        self.reply_queue.put(agent_message)
+
+
+@ray.remote
+class FileSystemActor():
+    """Processes messages prefixed with /fs. This actor will
+    send the message as input to a RetrieverQA chain, that
+    follows the Retrieval and Generation (RAG) tehnique to
+    query the documents from the index, and sends this context
+    to the LLM to generate the final reply.
+    """
+
+    def __init__(self, reply_queue: Queue):
+        self.reply_queue = reply_queue
+        index_actor = ray.get_actor(ACTOR_TYPE.READ.value)
+        handle = index_actor.get_index.remote()
+        vectorstore = ray.get(handle)
+        if not vectorstore:
+            return
+
+        self.chat_history = []
+        self.chat_provider = ConversationalRetrievalChain.from_llm(
+            OpenAI(temperature=0, verbose=True),
+            vectorstore.as_retriever()
+        )
+
+    def process_message(self, message: HumanChatMessage):
+        query = message.body.split(' ', 1)[-1]
+        
+        index_actor = ray.get_actor(ACTOR_TYPE.READ.value)
+        handle = index_actor.get_index.remote()
+        vectorstore = ray.get(handle)
```

Comment:
Not sure why we are reading this again? Let's read from disk directly in this actor and store as an instance attribute for reuse.

Reply:
The issue I faced was that when the index is updated with additional docs, the `index` reference from the index reference was not reflecting all docs. We have to investigate this more before we update. Will tackle this in the follow up PR.

Reply:
Sounds good